### ISSUE TITLE
Error: `Unexpected end of row` when setup `diesel`

### ISSUE THREAD

**Issue by kimyvgy** [diesel-rs/diesel#3579]
### Versions

- **Rust:** rustc 1.68.2 (9eb3afe9e 2023-03-27)
- **Diesel:** 2.0.3
- **Database:** SQLite
- **Operating System:** WSL2 - Ubuntu 18.04

### Feature Flags

- **diesel:** sqlite

## Problem Description

I got these errors when setup diesel:
```bash
> diesel migration run
Running migration 2023-03-30-181339_create_songs
Running migration 2023-03-31-021724_create_users
Unexpected end of row

> diesel print-schema
Unexpected end of row
```

The migrations are applied. I think the error is for `print-schema`.

The migrations:

```sql
-- 2023-03-30-181339_create_songs
CREATE TABLE songs (
  id INTEGER PRIMARY KEY,
  url VARCHAR(255) NOT NULL,
  status VARCHAR(10) NOT NULL DEFAULT 'queued',
  created_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
);

CREATE INDEX songs_status_idx ON songs (status);
CREATE INDEX songs_created_at_idx ON songs (created_at);

-- 2023-03-31-021724_create_users
CREATE TABLE users (
  id INTEGER PRIMARY KEY,
  username VARCHAR(255) NOT NULL UNIQUE,
  avatar VARCHAR(255)
);
```

**Comment by weiznich** [comment]
I cannot reproduce that locally. Can you provide details about your sqlite version?

**Comment by inferiorhumanorgans** [comment]
```
$ cargo tree -i libsqlite3-sys --depth 1
libsqlite3-sys v0.26.0
└── diesel v2.1.0
libsqlite3-sys v0.26.0

$ sqlite3 --version
sqlite3 --version
3.24.0 2018-06-04 14:10:15 95fbac39baaab1c3a84fdfc82ccb7f42398b2e92f18a2a57bce1d4a713cbaapl

$ uname -vr
uname -vr
18.7.0 Darwin Kernel Version 18.7.0: Tue Jun 22 19:37:08 PDT 2021; root:xnu-4903.278.70~1/RELEASE_X86_64

**Comment by inferiorhumanorgans** [comment]
Digging into this a little more it seems like there's some interaction with the system version of sqlite3 (at least for me) and using `sqlite-bundled` instead of `sqlite` appears to get things working.

**Comment by inferiorhumanorgans** [comment]
The problem is that the version of sqlite bundled with the version of OSX I'm using does not support `pragma table_xinfo` and the codepath for `pragma table_info` does not work.  `table_xinfo` returns an extra column and the `table_info` path expects that extra column.

**Comment by weiznich** [comment]
Thanks for investigating. These insights are really helpful to understand the underlying problem. I will try to implement a fix for that in the next weeks. I likely need some help to testing the fix as soon as it is ready. Can I just ping you from the relevant PR then?

**Comment by inferiorhumanorgans** [comment]
Sure

**Comment by weiznich** [comment]
I had some moments to look through some code and I think I know whats happening here:

The problematic code is here: https://github.com/diesel-rs/diesel/blob/master/diesel_cli/src/infer_schema_internals/sqlite.rs#L144

For newer sqlite version we call `PRAGMA table_xinfo(...)`, for older versions we call `PRAGMA table_info(...)`. The former returns one more column than the later call. The linked code (specifically the `Queryable` impl some lines below) assumes the larger column number from `table_xinfo(...)`

I would like to change the code so that:

* It uses `sql_query` instead `dsl::sql`
* The `Queryable` impl is replaced by a `QueryableByName` impl that just uses a default value if the additional field is not there.

**Comment by inferiorhumanorgans** [comment]
Yeah that's my understanding of what's happening.  I've not taken the time to go much further, but if the `hidden` column is not being used then it should be safe to simply use `table_info` regardless.  From the SQLite docs:

> This pragma returns one row for each column in the named table, including generated columns and hidden columns. The output has the same columns as for PRAGMA table_info plus a column, "hidden", whose value signifies a normal column (0), a dynamic or stored generated column (2 or 3), or a hidden column in a virtual table (1). The rows for which this field is non-zero are those omitted for PRAGMA table_info.

**Comment by weiznich** [comment]
It's necessary to use `table_xinfo` as we want to get information about generated columns as well.

**Comment by inferiorhumanorgans** [comment]
Ah, ok.  In that case, yes, using `QueryableByName` seems like a good idea to me.

**Comment by dsp** [comment]
@inferiorhumanorgans @kimyvgy can you give https://github.com/diesel-rs/diesel/pull/3713 a try?

**Comment by weiznich** [comment]
Fixed by #3713
